### PR TITLE
Basic Seldon Core Deployment Service

### PR DESCRIPTION
**Pull request by stefannica**
## Basic Seldon Core Deployment Service

This PR adds a Seldon Core Deployment Service as part of the new Seldon Core integration. In the absence of a Seldon Core Deployer stack component (TBD) and a Seldon Core Deployment Step (TBD), the Service needs to be configured and managed externally, e.g.:

```
from zenml.integrations.seldon.services import (
    SeldonDeploymentConfig,
    SeldonDeploymentService,
)

service_config = SeldonDeploymentConfig(
    model_uri="s3://zenfiles/seldon_model_deployer/output/374/seldon",
    model_name="mnist",
    implementation="SKLEARN_SERVER",
    secret_name="seldon-init-container-secret",
    pipeline_name=step_env.pipeline_name,
    pipeline_run_id=step_env.pipeline_run_id,
    pipeline_step_name=step_env.step_name,
    replicas=1,
    kubernetes_context="zenml-eks-sandbox",
    namespace="zenml-workloads",
    base_url="http://abb84c444c7804aa98fc8c097896479d-377673393.us-east-1.elb.amazonaws.com",
)

service = SeldonDeploymentService(config=service_config)
service.start(timeout=120)
```

The PR consists of the following:

1. a Seldon Core client library, which is probably the largest file in this PR. Implementing this could not have been avoided, as there currently is no Python client library for configuring and managing Seldon Core Kubernetes resources. The client library makes use of Pydantic objects to model the basic structure of a SeldonDeployment Kubernetes CRD and implements basic CRUD operations for these. It also implements some ZenML specific conventions to distinguish between SeldonDeployment instances that are managed by ZenML and those that are managed by other parties.

2. a SeldonDeploymentService that manages the life-cycle of a single remote SeldonDeployment. It implements provisioning and de-provisioning of SeldonDeployment resources, as well as tracking their operational status. Some configuration attributes, such as the Kubernetes context and namespace are only temporarily hosted by the Service configuration. They will be moved to the model deployer stack component in a follow-up PR that implements that component.

3. a continuous deployment example with a custom step implementation that facilitates the collection of all information required to create and run a SeldonDeploymentService as part of a pipeline. To run the example, Seldon Core must be installed in a Kubernetes cluster and a secret containing the authentication credentials required to access the Artifact Store has to be manually created in the `kubeflow` namespace (still TODO is to create this secret automatically from a ZenML Secret). The example features a full set of command line arguments that can be used to switch between tensorflow and sklearn as the type of model being trained, as well as to set different hyperparameter values for the training. The custom continuous deployment step implementation serves as s baseline for the built-in Seldon ZenML step that will be implemented in a follow-up PR (along with the SeldonDeployer stack component).

4. some bug fixes required to test this PR in an AWS stack are also included:

  a. fix S3 fileio bugs with directory copying
  b. fixed a bug with the secret manager (if no secrets are configured, the Kubeflow pipeline deployment fails)
  c. prevent MLMD database migration with kubeflow metadata store

### TODO (this PR)

- [x] add example README
- [x] use the implicit (in-pod) Kubernetes credentials to make it work with Kubeflow, when Kubeflow and Seldon Core are installed in the same cluster

### TODO (follow-up PR)

- [ ] Seldon Core model deployer as a stack component that holds the global configuration
- [ ] standard Seldon Core deployer step
- [ ] documentation section for Seldon Core
- [ ] use a ZenML Secret to configure credentials
- [ ] auto-detect the Seldon implementation (e.g. tensorflow/sklearn) from the model artifact data type

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/features/integrations) table.
- [ ] I have added tests to cover my changes.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: examples/seldon_deployment/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+    logger.info(
+        f"Loading last service deployed by step {step_name} and "
+        f"pipeline {step_env.pipeline_name}..."
+    )
+
+    try:
+        service = cast(
+            SeldonDeploymentService,
+            # TODO [HIGH]: catch errors that are raised because a previous step used
+            #   an integration that is no longer available
```
Comment (AlexejPenner) [inline examples/seldon_deployment/pipeline.py:219]: Update_todo gh action only checks at 'src/zenml' and at 'tests' for these Todos. If this isn't doable this sprint, you should create the ticket in jira manually.

File: examples/seldon_deployment/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+            if not fileio.file_exists(model.uri):
+                raise RuntimeError(
+                    f"Expected sklearn model artifact was not found at {model_uri}"
+                )
+            fileio.copy(
+                model_uri, os.path.join(served_model_uri, "model.joblib")
+            )
+        else:
+            # TODO [MEDIUM]: implement model preprocessing for other built-in
+            #   Seldon server implementations
```
Comment (AlexejPenner) [inline examples/seldon_deployment/pipeline.py:282]: Same as above

File: examples/seldon_deployment/run.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
+#  Copyright (c) ZenML GmbH 2022. All Rights Reserved.
```
Comment (AlexejPenner) [inline examples/seldon_deployment/run.py:1]: The Readme for this example is still WIP?

File: examples/seldon_deployment/run.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    normalizer,
+    prediction_service_loader,
+    predictor,
+    seldon_model_deployer,
+    sklearn_evaluator,
+    sklearn_predict_preprocessor,
+    sklearn_trainer,
+    tf_evaluator,
+    tf_predict_preprocessor,
+    tf_trainer,
```
Comment (AlexejPenner) [inline examples/seldon_deployment/run.py:35]: While its quite awesome that you have implementations for sklearn as well as tensorflow, it adds complexity to this example. Why do you think its necessary to show off implementations for two model types?

File: examples/seldon_deployment/run.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        running=True,
+    )
+    if service:
+        print(
+            f"The Seldon prediction server is running remotely as a Kubernetes "
+            f"service and accepts inference requests at:\n"
+            f"    {service.prediction_url}\n"
+            f"To stop the service, re-run the same command and supply the "
+            f"`--stop-service` argument."
+        )
```
Comment (AlexejPenner) [inline examples/seldon_deployment/run.py:205]: What would you think about separating deployment and inference into two functions that can be called separately from the cli. It makes it more interactive for the user and makes it a bit more obvious for the user

File: src/zenml/integrations/seldon/seldon_client.py [hunk 6]
```diff
@@ -0,0 +1,735 @@
...
+
+    UNKNOWN_TYPE = "UNKNOWN_TYPE"
+    ROUTER = "ROUTER"
+    COMBINER = "COMBINER"
+    MODEL = "MODEL"
+    TRANSFORMER = "TRANSFORMER"
+    OUTPUT_TRANSFORMER = "OUTPUT_TRANSFORMER"
+
+
+class SedonDeploymentPredictiveUnit(BaseModel):
```
Comment (AlexejPenner) [inline src/zenml/integrations/seldon/seldon_client.py:57]: Typo in classname: Se**l**donDeploymentPredictiveUnit

File: src/zenml/integrations/seldon/seldon_client.py [hunk 7]
```diff
@@ -0,0 +1,735 @@
+#  Copyright (c) ZenML GmbH 2022. All Rights Reserved.
```
Comment (AlexejPenner) [inline src/zenml/integrations/seldon/seldon_client.py:1]: I believe all the classes within this file could benefit from descriptive docstring that helps understand what the class does/represents. The docstrings that some of the classes already have are perfect :)

File: src/zenml/integrations/seldon/seldon_client.py [hunk 8]
```diff
@@ -0,0 +1,735 @@
...
+    class Config:
+        """Pydantic configuration class."""
+
+        # validate attribute assignments
+        validate_assignment = True
+        # Ignore extra attributes from the CRD that are not reflected here
+        extra = "ignore"
+
+
+class SedonDeploymentPredictor(BaseModel):
```
Comment (AlexejPenner) [inline src/zenml/integrations/seldon/seldon_client.py:80]: Same typo

File: src/zenml/integrations/seldon/services/seldon_deployment.py [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+
+        Args:
+            request: a numpy array representing the request
+
+        Returns:
+            A numpy array representing the prediction returned by the service.
+        """
+        if not self.is_running:
+            raise Exception(
+                "MLflow prediction service is not running. "
```
Comment (AlexejPenner) [inline src/zenml/integrations/seldon/services/seldon_deployment.py:270]:
```suggestion
                "Seldon prediction service is not running. "
```

File: src/zenml/integrations/seldon/services/seldon_deployment.py [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+
+    def predict(self, request: "NDArray[Any]") -> "NDArray[Any]":
+        """Make a prediction using the service.
+
+        Args:
+            request: a numpy array representing the request
+
+        Returns:
+            A numpy array representing the prediction returned by the service.
+        """
```
Comment (AlexejPenner) [inline src/zenml/integrations/seldon/services/seldon_deployment.py:267]: Does this mean we will only support prediction models that take numpy arrays as input?

**Review by AlexejPenner — CHANGES_REQUESTED** [review]
What a PR, really looking forward to reading the examples README to see what the user needs to do to get all this goodness. Here's some small comments, questions and CRs

File: src/zenml/integrations/seldon/seldon_client.py (hunk 6 above)
Comment (stefannica) [reply src/zenml/integrations/seldon/seldon_client.py:57]: Good catch ! Thanks !

File: src/zenml/integrations/seldon/services/seldon_deployment.py (hunk 10 above)
Comment (stefannica) [reply src/zenml/integrations/seldon/services/seldon_deployment.py:267]: No. The Seldon Core prediction API [supports several data types](https://docs.seldon.io/projects/seldon-core/en/latest/reference/apis/external-prediction.html#rest-api). We only offer this predict function as a convenience that can be used to contact the deployment service and run a quick inference request. Implementing a full prediction client is out of scope for this integration. Besides, there is an entire seldon-core Python library dedicated to this.

File: examples/seldon_deployment/run.py (hunk 4 above)
Comment (stefannica) [reply examples/seldon_deployment/run.py:35]: An important side of the Seldon Core integration is being able to deploy several model types for online prediction. Sklearn and tensorflow are both here to show that we're flexible on that end and that no code changes are required in the pipeline to accommodate this. Auto-detecting the deployer implementation from the model type is also a nice feature that I want to add to this PR.

File: examples/seldon_deployment/run.py (hunk 3 above)
Comment (stefannica) [reply examples/seldon_deployment/run.py:1]: Yes, you're correct. Next on my list.

File: examples/seldon_deployment/pipeline.py (hunk 1 above)
Comment (stefannica) [reply examples/seldon_deployment/pipeline.py:219]: I want to remake this step into a standard Seldon Core step in my next PR (it's part of the Seldon Core integration story). The TODOs will be reflected there as well, so I'll just leave them here for now, if that's ok with you.

File: examples/seldon_deployment/run.py (hunk 5 above)
Comment (stefannica) [reply examples/seldon_deployment/run.py:205]: This is a great idea. However, I spent too much time as it is on pimping up this example and there are tons of other things I need to do to finish up the integration (seldon core deployer component, secret management, maybe add in some endpoint health-checks). I would rather wave this idea for another time.

Comment (AlexejPenner) [reply examples/seldon_deployment/run.py:205]: Alright leave this as is for now, If I get the time I'll do it within the examples overhaul :)

Comment (stefannica) [reply examples/seldon_deployment/run.py:205]: Scratch that, I went ahead and implemented it anyway :D

**Comment by stefannica** [comment]
@AlexejPenner I pushed a couple of commits that hopefully address all your comments. I also rebased the PR on top of `devel` to incorporate the latest changes into it.

File: src/zenml/integrations/seldon/services/seldon_deployment.py [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+        pipeline_name: the name of the pipeline that was used to deploy the
+            model
+        pipeline_run_id: the ID of the pipeline run that deployed the model
+        pipeline_step_name: the name of the pipeline step that deployed the
+            model
+        replicas: number of replicas to use for the prediction service
+        model_metadata: optional model metadata information (see
+            https://docs.seldon.io/projects/seldon-core/en/latest/reference/apis/metadata.html)
+        extra_args: additional arguments to pass to the Seldon Core deployment
+            resource configuration
```
Comment (strickvl) [inline src/zenml/integrations/seldon/services/seldon_deployment.py:56]:
```suggestion
    Attributes:
        model_uri: URI of the model (or models) to serve.
        model_name: the name of the model. Multiple versions of the same model
            should use the same model name.
        model_format: the format of the model being served.
        implementation: the Seldon Core implementation used to serve the model.
        pipeline_name: the name of the pipeline that was used to deploy the
            model.
        pipeline_run_id: the ID of the pipeline run that deployed the model.
        pipeline_step_name: the name of the pipeline step that deployed the
            model.
        replicas: number of replicas to use for the prediction service.
        model_metadata: optional model metadata information (see
            https://docs.seldon.io/projects/seldon-core/en/latest/reference/apis/metadata.html).
        extra_args: additional arguments to pass to the Seldon Core deployment
            resource configuration.
```

File: src/zenml/integrations/seldon/services/seldon_deployment.py [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+    """Seldon Core deployment service status."""
+
+
+class SeldonDeploymentService(BaseService):
+    """A service that represents a Seldon Core deployment server.
+
+
+    Attributes:
+        config: service configuration
+        status: service status
```
Comment (strickvl) [inline src/zenml/integrations/seldon/services/seldon_deployment.py:96]:
```suggestion
        config: service configuration.
        status: service status.
```

File: src/zenml/services/service.py [hunk 13]
```diff
@@ -263,6 +279,14 @@ def deprovision(self, force: bool = False) -> None:
             f"Deprovisioning resources not implemented for {self}."
         )
 
+    def update(self, config: ServiceConfig) -> None:
+        """Update the service configuration.
+
+        Args:
+            config: the new service configuration
```
Comment (strickvl) [inline src/zenml/services/service.py:286]:
```suggestion
            config: the new service configuration.
```

File: src/zenml/services/utils.py [hunk 14]
```diff
@@ -42,6 +42,9 @@ def load_last_service_from_step(
     Returns:
         A BaseService instance that represents the service or None if no service
         was created during the last execution of the pipeline step.
+
+    Raises:
+        KeyError: if the pipeline or step name is not found in the execution
```
Comment (strickvl) [inline src/zenml/services/utils.py:47]:
```suggestion
        KeyError: if the pipeline or step name is not found in the execution.
```

**Review by strickvl — CHANGES_REQUESTED** [review]
Just some tiny nits / consistency things. I'll have to spend more time with this to properly understand how it's implemented, but I did what I could :)

My bigger feedback on the example is that I think we might want to / (have to) find a way to make this comprehensible to people who aren't K8S-native / -fluent. It might just be me, but I found the setup instructions a bit overwhelming. I'm not sure how comprehensible it is to someone whose professional life / obligations mean they don't have much experience with these concepts, the commands or just in general the whole universe of how this piece of ZenML works.

So that people don't get left behind, I almost would want a "Kubernetes for dummies" section where there is a ton of from-basics explanation and context going on. Maybe it doesn't have to be in this example -- probably good in a video or a blog -- but at the very least there should be some hints and support here. When we tell people to set up an Istio service mesh, we should explain what it does, why we're doing this etc. At the moment I think we might lose some people along the way.

File: examples/seldon_deployment/README.md [hunk 15]
```diff
@@ -0,0 +1,397 @@
...
+## Overview
+
+The example uses the
+[Fashion-MNIST](https://github.com/zalandoresearch/fashion-mnist) dataset to
+train a classifier using either [Tensorflow (Keras)](https://www.tensorflow.org/)
+or [scikit-learn](https://scikit-learn.org/stable/). Different
+hyperparameter values (e.g. the number of epochs and learning rate for the Keras
+model, solver and penalty for the scikit-learn logical regression) can be
+supplied as command line arguments to the `run.py` Python script.
+
```
[the remark this answers is not in the document]
Comment (stefannica) [reply examples/seldon_deployment/README.md:31]: It's definitely doable, I already have some pics taken somewhere.

[the remark this answers is not in the document]
Comment (stefannica) [reply examples/seldon_deployment/README.md:31]: Done

**Comment by stefannica** [comment]
> Just some tiny nits / consistency things. I'll have to spend more time with this to properly understand how it's implemented, but I did what I could :)
> 
> My bigger feedback on the example is that I think we might want to / (have to) find a way to make this comprehensible to people who aren't K8S-native / -fluent. It might just be me, but I found the setup instructions a bit overwhelming. I'm not sure how comprehensible it is to someone whose professional life / obligations mean they don't have much experience with these concepts, the commands or just in general the whole universe of how this piece of ZenML works.
> 

This is where the DevOps engineer comes in. As a Data Scientist, you don't need to install these infrastructure tools yourself, they're handed to you by someone who knows how to deal with Kubernetes.

> So that people don't get left behind, I almost would want a "Kubernetes for dummies" section where there is a ton of from-basics explanation and context going on. Maybe it doesn't have to be in this example -- probably good in a video or a blog -- but at the very least there should be some hints and support here. When we tell people to set up an Istio service mesh, we should explain what it does, why we're doing this etc. At the moment I think we might lose some people along the way.
> 
> (Also want to recognise that it could have been a LOT more gnarly, and this integration already makes it a ton easier. I guess I'm just asking if it can be any more signposted / easier?)

I don't think this can be avoided. Production grade tools like Seldon Core have to be installed, properly configured and maintained and there is no way around it, unless they are provided as a managed service. The idea that this can be automated away and hidden behind some ZenML abstraction and that you don't need to know what they do and how they do it is a bit far fetched at the moment, in my humble opinion. Even Kubeflow is a pain to install and maintain. The fact that you didn't run across a similar complexity regarding Kubeflow is largely due to the fact that ZenML automates a local toy installation for it, but take my word for it, it's as beastly as it can get with complex Kubernetes apps. :)

**Review by AlexejPenner — APPROVED** [review]

**Review by strickvl — APPROVED** [review]

**Pull request merged**